api/authenticator.py

```diff
@@ -843,14 +852,44 @@ def authenticated_patron(self, _db, header):
                 # a registered SAMLAuthenticationProvider.
                 return provider
 
-            # Ask the SAMLAuthenticationProvider to turn its token
-            # into a Patron.
-            return provider.authenticated_patron(_db, provider_token)
+        if provider and provider_token:
+            # Turn the token/header into a patron
+            try:
+                if self._is_provider_available(_db, provider):
+                    return provider.authenticated_patron(_db, provider_token)
+                else:
+                    return REMOTE_INTEGRATION_FAILED
+            except (RequestTimedOut, RemoteIntegrationException) as ex:
+                name = provider.external_integration(_db).name
+                self.log.error(f"Authentication Provider Error ({name}): {ex}")
+                self._record_provider_error(_db, provider, ex)
+                return REMOTE_INTEGRATION_FAILED
+            except PatronAuthorizationFailedException:
+                return INVALID_CREDENTIALS
 
         # We were unable to determine what was going on with the
         # Authenticate header.
         return UNSUPPORTED_AUTHENTICATION_MECHANISM
 
+    def _record_provider_error(
+        self, _db, provider: "AuthenticationProvider", error: Exception
+    ) -> ExternalIntegrationError:
+        record, _ = create(
+            _db,
+            ExternalIntegrationError,
+            external_integration_id=provider.external_integration_id,
+            time=utc_now(),
+            error=str(error),
+        )
+        return record
+
+    def _is_provider_available(self, _db, provider):
+        if provider.external_integration(_db).status != ExternalIntegration.STATUS.RED:
+            return True
+        # Provider status is RED, should we still try a request?
+        # Let's sample 1 in 10 requests to keep the errors up to date
+        return random.random() >= 0.9
```

Comment:
Randomly chosen sample rate.

Reply:
I think this one might've been accidentally resolved, so unresolving. ☝🏽

Reply:
I removed this logic in this [commit](https://github.com/ThePalaceProject/circulation/pull/980/commits/d456f194e94fe86fc77828f81e0a883d50912ca7) hence marked it as resolved